In [2]:
!pip install transformers

# Install the datasets library
!pip install datasets

In [3]:
import torch
from huggingface_hub import login
from transformers import AutoTokenizer, AutoModelForCausalLM

login("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")

# Replace 'Qwen/Qwen2.5-1.5B-Instruct' with the model you want to download
model_name = "Qwen/Qwen2.5-1.5B-Instruct"
token = "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"  # Use your token here

# Load the tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name, token=token)

# Load the model
model = AutoModelForCausalLM.from_pretrained(model_name, token=token)

In [21]:
from datasets import load_dataset

# Load the GSM8K dataset
dataset = load_dataset("gsm8k", "main", split="test")
test_data = dataset
print(test_data[0])

{'question': ["Janet’s ducks lay 16 eggs per day. She eats three for breakfast every morning and bakes muffins for her friends every day with four. She sells the remainder at the farmers' market daily for $2 per fresh duck egg. How much in dollars does she make every day at the farmers' market?", 'A robe takes 2 bolts of blue fiber and half that much white fiber.  How many bolts in total does it take?'], 'answer': ['Janet sells 16 - 3 - 4 = <<16-3-4=9>>9 duck eggs a day.\nShe makes 9 * 2 = $<<9*2=18>>18 every day at the farmer’s market.\n#### 18', 'It takes 2/2=<<2/2=1>>1 bolt of white fiber\nSo the total amount of fabric is 2+1=<<2+1=3>>3 bolts of fabric\n#### 3']}


In [27]:
# Set the model to evaluation mode to save memory
model.eval()

# Solve each question in the test set
results = []
for i in range(5):  # Process the first 5 items as an example
    print(test_data[i]["question"])
    question = test_data[i]["question"]
    inputs = tokenizer(f"Q: {question}\nA:", return_tensors="pt")

    # Generate an answer
    with torch.no_grad():
        outputs = model.generate(**inputs, max_new_tokens=512)
    answer = tokenizer.decode(outputs[0], skip_special_tokens=True)

    # Save the question, generated answer, and reference answer
    results.append({
        "question": question,
        "model_answer": answer,
        "reference_answer": test_data[i]["answer"]
    })

Janet’s ducks lay 16 eggs per day. She eats three for breakfast every morning and bakes muffins for her friends every day with four. She sells the remainder at the farmers' market daily for $2 per fresh duck egg. How much in dollars does she make every day at the farmers' market?
A robe takes 2 bolts of blue fiber and half that much white fiber.  How many bolts in total does it take?
Josh decides to try flipping a house.  He buys a house for $80,000 and then puts in $50,000 in repairs.  This increased the value of the house by 150%.  How much profit did he make?
James decides to run 3 sprints 3 times a week.  He runs 60 meters each sprint.  How many total meters does he run a week?
Every day, Wendi feeds each of her chickens three cups of mixed chicken feed, containing seeds, mealworms and vegetables to help keep them healthy.  She gives the chickens their feed in three separate meals. In the morning, she gives her flock of chickens 15 cups of feed.  In the afternoon, she gives her chi

In [ ]:
for result in results[:5]:  # Show the first 5 results
    print("Question:", result["question"])
    print("\n")
    print("Model's Answer:", result["model_answer"])
    print("\n")
    print("Reference Answer:", result["reference_answer"])
    print("\n")
    print("---------")
    print("\n")

Question: Janet’s ducks lay 16 eggs per day. She eats three for breakfast every morning and bakes muffins for her friends every day with four. She sells the remainder at the farmers' market daily for $2 per fresh duck egg. How much in dollars does she make every day at the farmers' market?
Model's Answer: Q: Janet’s ducks lay 16 eggs per day. She eats three for breakfast every morning and bakes muffins for her friends every day with four. She sells the remainder at the farmers' market daily for $2 per fresh duck egg. How much in dollars does she make every day at the farmers' market?
A: If there are 30 days in a month, then in one day, Janet gets 16 * 3 = 48 eggs from her ducks to sell. From those, she eats 3 eggs for breakfast so that leaves 48 - 3 = 45 eggs for baking muffins. This means that 48 - 3 - 4 = 41 eggs remain for selling at the farmers’ market. At $2 per egg, this is an income of 41 * $2 / egg = $82.

Therefore, the answer is 82.
Reference Answer: Janet sells 16 - 3 - 4 = 

## Full Output

Question: Janet’s ducks lay 16 eggs per day. She eats three for breakfast every morning and bakes muffins for her friends every day with four. She sells the remainder at the farmers' market daily for $2 per fresh duck egg. How much in dollars does she make every day at the farmers' market?

Model's Answer: 
Q: Janet’s ducks lay 16 eggs per day. She eats three for breakfast every morning and bakes muffins for her friends every day with four. She sells the remainder at the farmers' market daily for $2 per fresh duck egg. How much in dollars does she make every day at the farmers' market?
A: If there are 30 days in a month, then in one day, Janet gets 16 * 3 = 48 eggs from her ducks to sell. From those, she eats 3 eggs for breakfast so that leaves 48 - 3 = 45 eggs for baking muffins. This means that 48 - 3 - 4 = 41 eggs remain for selling at the farmers’ market. At $2 per egg, this is an income of 41 * $2 / egg = $82.

Therefore, the answer is 82.

Reference Answer: Janet sells 16 - 3 - 4 = 9 duck eggs a day.
She makes 9 * 2 = 18 every day at the farmer’s market.
=> 18

---------

Question: A robe takes 2 bolts of blue fiber and half that much white fiber.  How many bolts in total does it take?

Model's Answer: 
Q: A robe takes 2 bolts of blue fiber and half that much white fiber.  How many bolts in total does it take?
A: First find the amount of white fiber the robe needs: 0.5 * 2 = 1 bolt. Then add the amount of blue fiber to find the total number of bolts needed: 1 + 2 = 3 bolts.

Working backwards:

The answer is 3.

Reference Answer: It takes 2/2=1 bolt of white fiber
So the total amount of fabric is 2+1=3 bolts of fabric
=> 3

---------


Question: Josh decides to try flipping a house.  He buys a house for $80,000 and then puts in $50,000 in repairs.  This increased the value of the house by 150%.  How much profit did he make?

Model's Answer: 
Q: Josh decides to try flipping a house.  He buys a house for $80,000 and then puts in $50,000 in repairs.  This increased the value of the house by 150%.  How much profit did he make?
A: The house is now worth 80000 + 50000 = $130,000. The value has gone up by 130000 * .15 = $19,000. So his profit was 130000 - 80000 - 50000 = $60,000. The answer is 60000.

Reference Answer: The cost of the house and repairs came out to 80,000+50,000=130,000
He increased the value of the house by 80,000*1.5=120,000
So the new value of the house is 120,000+80,000=200,000
So he made a profit of 200,000-130,000=70,000
=> 70000

---------

Question: James decides to run 3 sprints 3 times a week.  He runs 60 meters each sprint.  How many total meters does he run a week?

Model's Answer: 
Q: James decides to run 3 sprints 3 times a week.  He runs 60 meters each sprint.  How many total meters does he run a week?
A: To find the total distance James runs in a week, we need to multiply the number of sprints by the length of each sprint and then multiply that result by the number of times he runs per week.

First, calculate the distance for one complete session:
Distance per session = sprint length * number of sprints = 60 meters * 3 = 180  meters

Next, since James runs this session three times a week, we multiply the distance per session by the number of sessions:
Total distance per week = distance per session * number of sessions per week = 180 meters * 3 = 540  meters

So, the total distance James runs in a week is: 540 meters.

Reference Answer: He sprints 3*3=9 times
So he runs 9*60=540 meters
=> 540

---------

Question: Every day, Wendi feeds each of her chickens three cups of mixed chicken feed, containing seeds, mealworms and vegetables to help keep them healthy. She gives the chickens their feed in three separate meals. In the morning, she gives her flock of chickens 15 cups of feed.  In the afternoon, she gives her chickens another 25 cups of feed.  How many cups of feed does she need to give her chickens in the final meal of the day if the size of Wendi's flock is 20 chickens?

Model's Answer: 
Q: Every day, Wendi feeds each of her chickens three cups of mixed chicken feed, containing seeds, mealworms and vegetables to help keep them healthy. She gives the chickens their feed in three separate meals. In the morning, she gives her flock of chickens 15 cups of feed.  In the afternoon, she gives her chickens another 25 cups of feed. How many cups of feed does she need to give her chickens in the final meal of the day if the size of Wendi's flock is 20 chickens?
A: First find the total number of cups of feed given out: 15 cups + 25 cups = 40 cups
Then divide that amount by the number of chickens to find the amount of feed per chicken: 40 cups / 20 chickens = 2 cups per chicken
Now multiply the number of chickens by the amount of feed per chicken to find the total amount of feed: 2 cups/chicken * 20 chickens = 40 cups

Therefore, the answer is 40.

Reference Answer: If each chicken eats 3 cups of feed per day, then for 20 chickens they would need 3*20=<<3*20=60>>60 cups of feed per day.
If she feeds the flock 15 cups of feed in the morning, and 25 cups in the afternoon, then the final meal would require 60-15-25=20 cups of chicken feed.
=> 20

---------